### PR TITLE
DBZ-7300 Review PostgreSQL documentations with new Snapshotter interfaces

### PR DESCRIPTION
**Pull request by mfvitale**
Doc for https://github.com/debezium/debezium/pull/5157

### CODE REVIEW

**Comment by mfvitale** [comment]
Hi @roldanbob any chance to have a look to this? Thanks.

**Comment by roldanbob** [comment]
> Hi @roldanbob any chance to have a look to this? Thanks.

Sorry! In the rush to the 2.3.7 product release, I missed this! I'll look at it now.

File: documentation/modules/ROOT/pages/connectors/postgresql.adoc [hunk 1]
```diff
@@ -242,106 +242,164 @@ ifdef::community[]
...
+     * @return {@code true} if streaming should resume from the start of the snapshot
+     * transaction, or {@code false} for when a connector resumes and takes a snapshot,
      * streaming should resume from where streaming previously left off.
      */
     default boolean shouldStreamEventsStartingFromSnapshot() {
         return true;
     }
+
+    /**
+     * Lifecycle hook called once the snapshot phase is successful.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/connectors/postgresql.adoc:326]:
```suggestion
     * Lifecycle hook called after the snapshot phase is successful.
```

File: documentation/modules/ROOT/pages/connectors/postgresql.adoc [hunk 2]
```diff
@@ -242,106 +242,164 @@ ifdef::community[]
...
+     */
+    default void snapshotCompleted() {
+        // no operation
+    }
+
+    /**
+     * Lifecycle hook called once the snapshot phase is aborted.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/connectors/postgresql.adoc:333]:
```suggestion
     * Lifecycle hook called after the snapshot phase is aborted.
```

File: documentation/modules/ROOT/pages/connectors/postgresql.adoc [hunk 3]
```diff
@@ -242,106 +242,164 @@ ifdef::community[]
...
+        // no operation
+    }
+}
+----
+
+[source,java,indent=0,subs="+attributes"]
+.io.debezium.snapshot.spi.SnapshotQuery interface. All built-in snapshot query modes implement this interface.
+----
+/**
+ * {@link SnapshotQuery} is used to determine the query used during data snapshot
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/connectors/postgresql.adoc:345]:
```suggestion
 * {@link SnapshotQuery} is used to determine the query used during a data snapshot
```

File: documentation/modules/ROOT/pages/connectors/postgresql.adoc [hunk 4]
```diff
@@ -3284,15 +3342,55 @@ For example, +
 `exported` -  deprecated +
  +
 ifdef::community[]
-`custom` - The connector performs a snapshot according to the setting for the `snapshot.custom.class` property, which is a custom implementation of the `io.debezium.connector.postgresql.spi.Snapshotter` interface. +
+`custom` - The connector performs a snapshot according to the implementation provided via `snapshot.mode.custom.name` property, which is a custom implementation of the `io.debezium.spi.snapshot.Snapshotter` interface. +
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/connectors/postgresql.adoc:3345]:
```suggestion
`custom` - The connector performs a snapshot according to the implementation specified by the xref:postgresql-property-snapshot-mode-custom-name[`snapshot.mode.custom.name`] property, which defines a custom implementation of the `io.debezium.spi.snapshot.Snapshotter` interface. +
```

File: documentation/modules/ROOT/pages/connectors/postgresql.adoc [hunk 5]
```diff
@@ -3284,15 +3342,55 @@ For example, +
...
 endif::community[]
  +
 For more information, see the xref:#postgresql-connector-snapshot-mode-options[table of `snapshot.mode` options].
 
 ifdef::community[]
-|[[postgresql-property-snapshot-custom-class]]<<postgresql-property-snapshot-custom-class, `+snapshot.custom.class+`>>
+|[[postgresql-property-snapshot-mode-custom-name]]<<postgresql-property-snapshot-mode-custom-name, `+snapshot.mode.custom.name+`>>
+|No default
+| When `snapshot.mode` is set as `custom`, this setting must be set to specify the name of the custom implementation provided in the `name()` method defined by the 'io.debezium.spi.snapshot.Snapshotter' interface.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/connectors/postgresql.adoc:3353]:
```suggestion
| When `snapshot.mode` is set as `custom`, use this setting to specify the name of the custom implementation provided in the `name()` method that is defined by the 'io.debezium.spi.snapshot.Snapshotter' interface.
```

File: documentation/modules/ROOT/pages/connectors/postgresql.adoc [hunk 6]
```diff
@@ -3284,15 +3342,55 @@ For example, +
...
+The provided implementation is called on each app boot to determine whether to do a snapshot. See xref:postgresql-custom-snapshot[custom snapshotter SPI].
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/connectors/postgresql.adoc:3354]:
```suggestion
The provided implementation is called after a connector restart to determine whether to perform a snapshot. 
For more information, see xref:postgresql-custom-snapshot[custom snapshotter SPI].
```

File: documentation/modules/ROOT/pages/connectors/postgresql.adoc [hunk 7]
```diff
@@ -3284,15 +3342,55 @@ For example, +
...
+endif::community[]
+
+|[[postgresql-property-snapshot-locking-mode]]<<postgresql-property-snapshot-locking-mode, `+snapshot.locking.mode+`>>
+|`none`
+|Controls how the connector holds locks on tables while performing the schema snapshot: +
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/connectors/postgresql.adoc:3359]:
```suggestion
|Specifies how the connector holds locks on tables while performing a schema snapshot: +
Set one of the following options:
```

File: documentation/modules/ROOT/pages/connectors/postgresql.adoc [hunk 8]
```diff
@@ -3284,15 +3342,55 @@ For example, +
...
+ +
+`shared` -  The connector will hold a table lock that prevents exclusive table access for just the initial portion of the snapshot while the database schemas and other metadata are being read. The remaining work in a snapshot involves selecting all rows from each table, and this is done using a flashback query that requires no locks. +
+ +
+`none` -  The connector avoids locks entirely. This mode is only safe to use if no schema changes are happening while the snapshot is taken. +
+ +
+ifdef::community[]
+`custom` - The connector performs a snapshot according to the implementation provided via `snapshot.locking.mode.custom.name` property, which is a custom implementation of the `io.debezium.spi.snapshot.SnapshotLock` interface. +
+endif::community[]
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/connectors/postgresql.adoc:3367]:
```suggestion
`shared`:: The connector holds a table lock that prevents exclusive table access during the initial portion phase of the snapshot in which database schemas and other metadata are read. 
After the initial phase, the snapshot no longer requires table locks. 
 +
`none`:: The connector avoids locks entirely. +
+
[WARNING]
====
Do not use this mode if schema changes might occur during the snapshot. 
====
 
ifdef::community[]
`custom`:: The connector performs a snapshot according to the implementation specified by the xref:postgresql-property-snapshot-locking-mode-custom-name[`snapshot.locking.mode.custom.name`] property, which is a custom implementation of the `io.debezium.spi.snapshot.SnapshotLock` interface.
endif::community[]
```

File: documentation/modules/ROOT/pages/connectors/postgresql.adoc [hunk 9]
```diff
@@ -3284,15 +3342,55 @@ For example, +
...
+
+ifdef::community[]
+|[[postgresql-property-snapshot-locking-mode-custom-name]]<<postgresql-property-snapshot-locking-mode-custom-name, `+snapshot.locking.mode.custom.name+`>>
+|No default
+| When `snapshot.locking.mode` is set as `custom`, this setting must be set to specify the name of the custom implementation provided in the `name()` method defined by the 'io.debezium.spi.snapshot.SnapshotLock' interface.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/connectors/postgresql.adoc:3372]:
```suggestion
| When `snapshot.locking.mode` is set as `custom`, use this setting to specify the name of the custom implementation provided in the `name()` method that is defined by the 'io.debezium.spi.snapshot.SnapshotLock' interface.
```

File: documentation/modules/ROOT/pages/connectors/postgresql.adoc [hunk 10]
```diff
@@ -3284,15 +3342,55 @@ For example, +
...
+See xref:postgresql-custom-snapshot[custom snapshotter SPI].
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/connectors/postgresql.adoc:3373]:
```suggestion
For more information, see xref:postgresql-custom-snapshot[custom snapshotter SPI].
```

File: documentation/modules/ROOT/pages/connectors/postgresql.adoc [hunk 11]
```diff
@@ -3284,15 +3342,55 @@ For example, +
...
+|Controls how the connector queries data while performing the snapshot: +
+ +
+
+`select_all` - The connector performs a select all query. +
+ +
+ifdef::community[]
+`custom` - The connector performs a snapshot query according to the implementation provided via `snapshot.query.mode.custom.name` property, which is a custom implementation of the `io.debezium.spi.snapshot.SnapshotQuery` interface. +
+endif::community[]
+ +
+This is a more flexible alternative to xref:{context}-property-snapshot-select-statement-overrides[`snapshot.select.statement.overrides`]
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/connectors/postgresql.adoc:3387]:
```suggestion
|`select_all`
|Specifies how the connector queries data while performing a snapshot. +
Set one of the following options:

`select_all`:: The connector performs a `select all` query. 

ifdef::community[]
`custom`:: The connector performs a snapshot query according to the implementation specified by the xref:postgresql-property-snapshot-snapshot-query-mode-custom-name[`snapshot.query.mode.custom.name`] property, which defines a custom implementation of the `io.debezium.spi.snapshot.SnapshotQuery` interface. +
endif::community[]

This setting enables you to manage snapshot content in a more flexible manner compared to using the xref:postgresql-property-snapshot-select-statement-overrides[`snapshot.select.statement.overrides`] property.
```

File: documentation/modules/ROOT/pages/connectors/postgresql.adoc [hunk 12]
```diff
@@ -3284,15 +3342,55 @@ For example, +
...
+
+ifdef::community[]
+|[[postgresql-property-snapshot-snapshot-query-mode-custom-name]]<<postgresql-property-snapshot-query-mode-custom-name, `+snapshot.query.mode.custom.name+`>>
 |No default
-| A full Java class name that is an implementation of the `io.debezium.connector.postgresql.spi.Snapshotter` interface. Required when the `snapshot.mode` property is set to `custom`. See xref:postgresql-custom-snapshot[custom snapshotter SPI].
+| When `snapshot.query.mode` is set as `custom`, this setting must be set to specify the name of the custom implementation provided in the `name()` method defined by the 'io.debezium.spi.snapshot.SnapshotQuery' interface.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/connectors/postgresql.adoc:3392]:
```suggestion
| When xref:postgresql-property-snapshot-query-mode[`snapshot.query.mode`] is set as `custom`, use this setting to specify the name of the custom implementation provided in the `name()` method that is defined by the 'io.debezium.spi.snapshot.SnapshotQuery' interface.
```

File: documentation/modules/ROOT/pages/connectors/postgresql.adoc [hunk 13]
```diff
@@ -3284,15 +3342,55 @@ For example, +
...
+See xref:postgresql-custom-snapshot[custom snapshotter SPI].
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/connectors/postgresql.adoc:3393]:
```suggestion
For more information, see xref:postgresql-custom-snapshot[custom snapshotter SPI].
```

**Comment by mfvitale** [comment]
Thanks for the review! I addressed all your suggestions and reported also in the code.

**Pull request merged**